packages/langium/src/parser/langium-parser.ts

```diff
@@ -247,12 +248,19 @@ export class LangiumParser {
             this.stack.pop();
         }
         if (obj.$type === DatatypeSymbol) {
-            const node = obj.$cstNode;
-            return node.text;
+            return this.constructDataTypeResult(obj.$cstNode);
         }
         return obj;
     }
 
+    private constructDataTypeResult(node: CstNode): string {
+        let text = '';
+        for (const subNode of streamCst(node).filter(isLeafCstNode)) {
+            text += this.converter.convert(subNode.text, subNode);
+        }
+        return text;
```

Comment:
Is the datatype result sent to the value converter as well? This is necessary when the return type of a datatype rule is not string. Could you add a test for that, if there is none yet?

Example:
```
Number returns number:
    INT ('.' INT)?;

terminal INT returns number:
    /[0-9]+/;
```

Reply:
Thanks, I kind of forgot that data type rules should able be sent to the value converter! Fixed this and added tests for that case 👍